### PR TITLE
dialog: fix rtags of forking INVITE with 100rel

### PR DESCRIPTION
**Pull request by maximilianfridrich**
See https://github.com/baresip/baresip/issues/2114.

### CODE REVIEW

File: src/sip/dialog.c [hunk 1]
```diff
@@ -475,14 +477,46 @@ int sip_dialog_update(struct sip_dialog *dlg, const struct sip_msg *msg)
...
 
-	if (dlg->route.scheme.p == dlg->uri) {
+	err = pl_strdup(&rtag, msg->req ? &msg->from.tag : &msg->to.tag);
+	if (err)
+		goto out;
+
+	renc.end = 0;
+
+	err |= sip_msg_hdr_apply(msg, msg->req, SIP_HDR_RECORD_ROUTE,
+				 record_route_handler, &renc) ? ENOMEM : 0;
```
Comment (sreimers) [inline src/sip/dialog.c:495]: Thats looks a bit strange to me. If `record_route_handler` returns true, `ENOMEM` should returned?

Comment (maximilianfridrich) [reply src/sip/dialog.c:495]: Absolutely, thanks. I will still clean this up a bit, hence the draft status. I wanted to make use of the GitHub workflows to make sure the test I wrote also works on other platforms.

File: test/sipsess.c [hunk 2]
```diff
@@ -374,6 +419,127 @@ static void close_handler(int err, const struct sip_msg *msg, void *arg)
...
+			test->progr_ret_code = err;
+			goto out;
+		}
+
+		err = sipsess_set_prack_handler(test->c, prack_handler_fork_c);
+		if (err)
+			abort_test(test, err);
+
+		if (err)
+			mem_deref(desc);
```
Comment (cspiel1) [inline test/sipsess.c:493]: cleanup err handling

File: test/sipsess.c [hunk 3]
```diff
@@ -374,6 +419,127 @@ static void close_handler(int err, const struct sip_msg *msg, void *arg)
...
+
+	}
+
+	if (test->conn_action & CONN_PROGR_ANS) {
+		err = sipsess_answer(test->b, 200, "Answering", NULL, NULL);
+		if (err != test->answ_ret_code) {
```
Comment (cspiel1) [inline test/sipsess.c:499]: should we rename `answ_ret_code` --> `answ_err` ?

File: test/sipsess.c [hunk 4]
```diff
@@ -335,8 +346,42 @@ static void prack_handler(const struct sip_msg *msg, void *arg)
...
+static void prack_handler_fork_b(const struct sip_msg *msg, void *arg)
+{
+	int err;
+	struct test *test = arg;
+	(void)msg;
+
+	if (test->sdp_state == ANSWER_RECEIVED)
+		test->sdp_state = EARLY_CONFIRMED;
+
+	if ((err = pl_strcmp(&msg->to.tag, test->rtag_b)))
```
Comment (cspiel1) [inline test/sipsess.c:363]: I prefer two lines with a simple `if (err)`

**Comment by maximilianfridrich** [comment]
I had to remove the new test for retest as I was not able to write a test that would properly simulate a forked INVITE. Instead, I tested it with a SIPP scenario and the PRACKs now have the correct tags.

Other than that the behavior of re stays the same regarding forked INVITES. I.e. only the first 2XX response is ACKed, subsequent ones are ignored.

**Pull request merged**